### ISSUE TITLE
Matrix-matrix solver on GPU

### ISSUE THREAD

**Issue by blegouix** [ginkgo-project/ginkgo#1381]
Hello,

I want to implement a matrix matrix solver AX=B with A sparse and X and B dense matrixes. To do so, I tried two strategies:

- On the develop branch, use the classes Csr<>, Dense<> and Bicgstab<> with GPU executor.
- On the batch-develop branch, use the classes BatchCsr<>, BatchDense<> and BatchBicgstab<>. In this case, I perform previously a duplication of the matrix A using the BatchCsr constructor which permits the duplication.

I obtain the following behaviours:

![batch_compare_ginkgo](https://github.com/ginkgo-project/ginkgo/assets/22033714/d1fff72c-027c-4162-bb8e-d8130c87abe0)

The first solution is not satisfying because the columns of B and X are processed sequentialy in n_batch 1-block kernels (no parallelization of the individual problems). The second solution is not satisfying too because it involves a huge DtoD memorycopy and fills the memory with duplicated information.

Is there already a way to do it or is the developments related to new MultiVector class appropriate for this problem and should we expect a solver anytime soon ?

Any known workaround with for loop or Kokkos::parallel_for ?

**Comment by tcojean** [comment]
Hi Baptiste,

To get a bit more info, what is the size of your A matrix in this case?

The MultiVector is simply a rewrite of the current batch code (which uses Dense currently). We are trying to merge the entire batch functionality into our code.

**Comment by MarcelKoch** [comment]
Hi @blegouix, could you also specify how you are executing strategy 1? Our solvers can handle dense matrices as right-hand-side, so you can just call `solver.apply(...)` once.
But the batched solve might still be more efficient for small a `A`.

**Comment by pratikvn** [comment]
I think strategy 1 is the right choice for this problem as you have only 1 matrix. I assume n=10 and n_batch=1000 in the code above ? And you are calling the `solver->generate(A_mat)->apply(...)` only once and not within a loop ?

**Comment by blegouix** [comment]
Yes I call it only once as in the given code with n=10 and n_batch=1000. It is not the expected behavior ?

**Comment by pratikvn** [comment]
Yes, that is the expected behaviour. It is a bit weird that you are seeing 1000 separate kernel calls. Our monolithic solvers support multiple RHS, so there should be only one solver call. There might be multiple kernel calls inside that single solver call though (dot, norm, axpy, spmv etc).

Essentially, you do these solver operations for all rhs in parallel.

**Comment by blegouix** [comment]
Ok great, so I have to figure out what is causing that. It may be the solver/preconditionner choice ? I would be happy to take your ideas

**Comment by MarcelKoch** [comment]
You could add a `Convergence` logger to your solver to get how many iterations you are actually doing. If you need 1000 iterations for a 10x10 matrix, something seems to be wrong. Perhaps your matrix is actually singluar and thus the solver is not converging at all.

**Comment by blegouix** [comment]
I will add the Convergence logger but I can confirm that the solver converges (I tried with A=Identity or simple matrices like that and I obtain right result, plus error computed norm(B-AX) checked).

And I confirm that I have n_batch individual kernels launched in sequential (1000 is not the number of iterations). I tried with n_batch=10 or n_batch=100 and I counted them.

**Comment by blegouix** [comment]
Ok, this is related to the size of the matrix 10x10 and the use of a Jacobi-block preconditionner (which is obviously not the good choice). What I see is not the actual Bicgstab iteration but something related to the preconditionning, because 10 being smaller than 32, block-jacobi is enough to inverse the matrix by itself and solves directly the systems.

Great thanks,
